trl/trainer/accelerate_ppo.py

```diff
@@ -0,0 +1,542 @@
+        # check: https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html
+        # or: https://discuss.pytorch.org/t/use-distributed-data-parallel-correctly/82500/11
+        self.is_distributed = self.accelerator.distributed_type == "MULTI_GPU"
+
+        # init wandb on the main process:
+        if self.accelerator.is_main_process:
+            wandb.init(name='run-42', project='gpt2-test', config=config)
+            wandb.watch(self.model, log='all')
+
+    def _build_dataset(self, dataset_name: str="imdb"):
+        """
+        Build dataset for training. This builds the dataset from `load_dataset`, one should 
+        customize this function to train the model on its own dataset.
+        
+        Args:
+            dataset_name (`str`): 
+                The name of the dataset to be loaded.
+        """
+        # load imdb with datasets
+        ds = load_dataset(dataset_name, split='train')
+        ds = ds.rename_columns({'text': 'review', 'label': 'sentiment'})
+        ds = ds.filter(lambda x: len(x["review"])>200, batched=False)
+
+
+        self.input_size = LengthSampler(self.config["txt_in_min_len"], self.config["txt_in_max_len"])
+        self.output_size = LengthSampler(self.config["txt_out_min_len"], self.config["txt_out_max_len"])
+
+        def tokenize(sample):
+            sample["tokens"] = self.tokenizer.encode(sample["review"])[:self.input_size()]
+            sample["query"] = self.tokenizer.decode(sample["tokens"])
+            return sample
+
+        ds = ds.map(tokenize, batched=False)
+
+        def collater(data):
+            return dict((key, [d[key] for d in data]) for key in data[0])
+
+        self.dataloader = torch.utils.data.DataLoader(ds, batch_size=self.config['batch_size'], collate_fn=collater)
+
+    def get_response(self, query_tensors: torch.Tensor, **gen_kwargs):
+        """
+        Generate response given query.
+
+
+        Args:
+            query_tensors (`torch.LongTensor`): 
+                A tensor of shape (`batch_size`, `seq_len`) containing query tokens.
+            gen_kwargs (dict[str, Any]): 
+                Keyword arguments for generation.
+        
+        Returns: 
+            response_tensors (`torch.LongTensor`): 
+                A tensor of shape (`batch_size`, `gen_len`) containing response tokens. `gen_len` 
+                is the length of the generated response that is sampled from `LengthSampler`.
+        """
+        response_tensors = []
+        for i in range(self.config['batch_size']):
+            gen_len = self.output_size()
+
+            # In a multi-GPU setup the model is wrapped inside a DistributedDataParallel object.
+            # this means that the model needs to be called with the module attribute.
+            if self.accelerator.distributed_type == "MULTI_GPU":
+                response = self.model.module.generate(query_tensors[i].unsqueeze(dim=0),
```

Comment:
With that setup all the workers run the same generation - no speedup. We can leave it like that for now and add parallel generation in a next step.